### PR TITLE
Add u32 ops execution tests to the processor

### PR DESCRIPTION
**Pull request by grjte**
This PR adds following to the processor u32 operations tests file:
1. conversions and tests: manual and randomized tests
2. arithmetic operations: stubs out manual tests
3. bitwise operations: manual and randomized tests
4. comparison operations: adds manual + randomized tests for ```u32eq```; stubs out manual tests for the rest

It also addresses the u32split bug #70 .

A few points / questions:
- **Stubbed out tests**: let me know if you want me to remove these. I did that first, before realizing that many of the ops which looked implemented are actually not fully to spec yet. At the moment, unimplemented ones are ignored with an "unimplemented" message
- **Comparison operations tests for EQ**: I wrote these before realizing that the comparison operations are not fully implemented to spec. I stopped writing comparison tests at that point, but left these ones & ignored the tests of unimplemented cases with an "unimplemented" message
- **Impl / spec mismatch for bitwise ops:** should ```shl```, ```shr```, ```rotl```, ```rotr``` actually fail for b (bit shift) = 0? It’s not in the spec, but since it's written that way I did write that into the tests for now. Not sure that's really correct, though.
- **Failure handling (panic):** I handled this using ```catch_unwind``` so we can test multiple execution errors in the same test, for example in cases where we want to test 2 inputs to make sure they're both required to be valid. That approach was discussed [here](https://github.com/maticnetwork/miden/issues/43) and seemed like a better fit for these tests, since doing exact error matching for some errors (like FailedAssertion) requires specifying the step. Let me know if you prefer something else.
- **add ```proptest-regressions``` failure persistance files to source control?** The proptest book says these should be added so they'll be replayed before new cases are generated, but I see you haven't done this in winterfell (but it's not in the .gitignore either), so maybe it's not your preference. For now, I left it out of both places as well.
- As discussed, I fixed some failures, but left failing tests for the ops in the 2 open issues ([here](https://github.com/maticnetwork/miden/issues/71) and [here](https://github.com/maticnetwork/miden/issues/74))

### CODE REVIEW

File: processor/src/tests/u32_ops.rs [hunk 1]
```diff
@@ -0,0 +1,1009 @@
...
+    test_execution_failure(asm_op, &[U32_BOUND], "FailedAssertion");
+}
+
+/// This helper function tests a provided u32 bitwise assembly operation which takes a parameter
+/// to ensure that it fails when that parameter is out of bounds above the maximum.
+fn test_param_out_of_bounds(asm_op_base: &str, over: u64) {
+    let build_asm_op = |param: u64| format!("{}.{}", asm_op_base, param);
+
+    // --- test b > max ---------------------------------------------------------------------------
+    test_compilation_failure(build_asm_op(over).as_str(), "parameter value");
```
Comment (bobbinth) [inline processor/src/tests/u32_ops.rs:66]: I wonder if most (all?) of these helper functions should be in the module one level higher. I'd imagine they'd be useful for other tests as well.

File: processor/src/tests/u32_ops.rs [hunk 2]
```diff
@@ -0,0 +1,1009 @@
...
+/// rotated) and b (the amount of rotation).
+fn get_rotation_result(op_base: &str, a: u32, b: u32) -> u64 {
+    let rot_inverse = 32 - b;
+    match op_base {
+        "u32rotl" => {
+            // shift everything by b to get the upper bits of the result
+            let new_upper_bits = (a << b) as u32;
+            // move b bits which fall off the top during left shift to the result's low bits
+            let new_lower_bits = a >> rot_inverse;
+            (new_upper_bits | new_lower_bits) as u64
```
Comment (bobbinth) [inline processor/src/tests/u32_ops.rs:80]: Can't we use the built in [u32::rotate_left()](https://doc.rust-lang.org/std/primitive.u32.html#method.rotate_left) function here?

File: processor/src/tests/u32_ops.rs [hunk 3]
```diff
@@ -0,0 +1,1009 @@
...
+        }
+        "u32rotr" => {
+            // shift everything by b to get the lower bits of the result
+            let new_lower_bits = (a >> b) as u32;
+            // move b bits which fall off the bottom during right shift to the result's upper bits
+            let new_upper_bits = a << rot_inverse;
+            (new_upper_bits | new_lower_bits) as u64
```
Comment (bobbinth) [inline processor/src/tests/u32_ops.rs:87]: Can't we use the built in [u32::rotate_right()](https://doc.rust-lang.org/std/primitive.u32.html#method.rotate_right) function here?

**Review by bobbinth — CHANGES_REQUESTED** [review]
Looks great! A couple of minor comments in-line.

To answer the questions:
* Let's keep the stubbed out tests in there. I think it is helpful to have them.
* Shifts/rotations by 0: I think these should be possible. Could you create an issue to fix this in the assembly op parser?
* `catch_unwind` works! Thank you for putting together helper functions for this.
* Let's not check-in proptest regressions into the source control. And it is probably a good idea to formalize this via `.gitignore`.

File: processor/src/tests/u32_ops.rs (hunk 1 above)
Comment (grjte) [reply processor/src/tests/u32_ops.rs]: I moved all but one up a level & updated the ```execute``` import for the other test files. The other tests could probably be updated to use the new helpers (especially ```test_execution```), but I left that alone for now.

**Comment by grjte** [comment]
> Shifts/rotations by 0: I think these should be possible. Could you create an issue to fix this in the assembly op parser?
Issue created: #76 

Thanks for the pointer on the Rust built in rotation functions! Rust doesn't disappoint :)

File: processor/src/tests/u32_ops.rs (hunk 1 above)
Comment (bobbinth) [reply processor/src/tests/u32_ops.rs]: Thank you! Agreed - propagating `test_execution` should be done in another PR.

File: processor/src/tests/u32_ops.rs [hunk 4]
```diff
@@ -0,0 +1,950 @@
...
+            // should fail when b is out of bounds and provided via the stack
+            test_execution_failure(asm_opcode, &values, "FailedAssertion");
+        } else {
+            // should test for equality
+            let expected = if a==b { 1 } else { 0 };
+            // b provided via the stack
+            test_execution(asm_opcode, &values, &[expected]);
+            // b provided as a parameter
+            test_execution(format!("{}.{}", asm_opcode, b).as_str(), &[a], &[expected]);
+        }
```
Comment (bobbinth) [inline processor/src/tests/u32_ops.rs:939]: I'm curious if proptest hits this cases regularly. If it sampled u64 values uniformly at random, these would be hit very rarely. But I believe it is actually smarter than that. So, just want to confirm.

File: processor/src/tests/u32_ops.rs [hunk 5]
```diff
@@ -0,0 +1,950 @@
...
+    unimplemented!();
+}
+
+// U32 OPERATIONS TESTS - RANDOMIZED
+// ================================================================================================
+
+// This is a proptest strategy for generating a random word of u64 values.
+fn rand_word() -> impl Strategy<Value = Vec<u64>> {
+    prop::collection::vec(any::<u64>(), WORD_LEN)
+}
```
Comment (bobbinth) [inline processor/src/tests/u32_ops.rs:674]: Should this also be a more general helper (at the root for test module)?

**Review by bobbinth — CHANGES_REQUESTED** [review]
Looks great! I left a couple of small questions inline. Also, one question:

There are a few tests which use the "old" approach (don't make use of `test_execution()` helper). For example: `u32test_proptest()`, `u32cast_proptest()` etc. It wasn't immediately obvious to me why these can't also rely on the new approach. Is there a reason for this?

File: processor/src/tests/u32_ops.rs (hunk 4 above)
Comment (grjte) [reply processor/src/tests/u32_ops.rs:939]: I wanted to test it with a couple runs, but this test is currently failing because the comparison ops aren't implemented to spec yet.

However - I'm not sure how much smarter proptest actually is - it's great for finding the minimal failing input once it's got an error, but for the initial values I think it may just be sampling uniformly at random as you said. I looked into this a bit with the last PR that I used proptest on, because it was missing an important edge case and I was wondering if a check I was doing inside the proptest was even needed. I ended up implementing the edge case manually, but I left the check inside the proptest because I didn't want a false positive.

From their crate:

>Given infinite time, property testing will eventually explore the whole input space to a test. However, time is not infinite, so only a randomly sampled portion of the input space can be explored. This means that property testing is extremely unlikely to find single-value edge cases in a large space.

**Comment by grjte** [comment]
>There are a few tests which use the "old" approach (don't make use of test_execution() helper). For example: u32test_proptest(), u32cast_proptest() etc. It wasn't immediately obvious to me why these can't also rely on the new approach. Is there a reason for this?

Not a good one :) I think I just missed them somehow when I refactored the rest. Thanks!

For this update I changed u32testw (manual test) and the conversions and tests proptests (as well as lifting the proptest helper you mentioned)

File: processor/src/tests/u32_ops.rs (hunk 4 above)
Comment (bobbinth) [reply processor/src/tests/u32_ops.rs:890]: Should we then sample from a `u32` space for the "else" clause?

Comment (grjte) [reply processor/src/tests/u32_ops.rs:890]: Ahh fair point. In that case, is it just better to do 2 tests - one that samples u32 for the valid case and one that samples u64 for failing & maybe even filters out the u32s? (Although it seems like filter may not be worth the cost. Opinion on that?)

Comment (bobbinth) [reply processor/src/tests/u32_ops.rs:890]: Maybe invalid cases can be handled in manual tests? And the test validity of operations over valid inputs using randomized tests?

**Review by bobbinth — APPROVED** [review]
